src/app/qgisapp.cpp

```diff
@@ -6583,14 +6598,22 @@ bool QgisApp::addProject( const QString &projectFile )
     QgsSettings settings;
 
 #ifdef WITH_BINDINGS
-    // does the project have any macros?
     if ( mPythonUtils && mPythonUtils->isEnabled() )
     {
+      // does the project have any macros?
       if ( !QgsProject::instance()->readEntry( QStringLiteral( "Macros" ), QStringLiteral( "/pythonCode" ), QString() ).isEmpty() )
       {
         auto lambda = []() {QgisApp::instance()->enableProjectMacros();};
         QgsGui::pythonMacroAllowed( lambda, mInfoBar );
       }
+
+      // does the project have expression functions?
+      const QString projectFunctions = QgsProject::instance()->readEntry( QStringLiteral( "ExpressionFunctions" ), QStringLiteral( "/pythonCode" ), QString() );
+      if ( !projectFunctions.isEmpty() )
+      {
+        auto lambda = []( const QString & pythonProjectFunctions ) {QgsPythonRunner::run( pythonProjectFunctions );};
```

Comment:
If this logic lives in app, how will projects published to server handle this?

Reply:
Right, we've moved it to `QgsExpressions` so that it can be used by QGIS Server as well.

Reply:
@gacarrillor , @nyalldawson , I propose an alternative approach to adding this into QgsExpression (which would probably have been better suited in QgsExpressionUtils anyways ;) ). 

I think this should happen in QgsProject::read (and cleared in QgsProject::clear). That way, we not only gain support for QGIS server, but also and most importantly qgis_process.

Reply:
Implemented in https://github.com/qgis/QGIS/pull/58354/commits/c5286e4dcc86dd78f445e31dd70006240c1ef935